lib/resources/entities.js

```diff
@@ -78,21 +78,34 @@ module.exports = (service, endpoint) => {
 
   }));
 
+  // Create a single entity or bulk create multiple entities.
+  // In either case, this appends new entities to a dataset.
   service.post('/projects/:id/datasets/:name/entities', endpoint(async ({ Datasets, Entities }, { auth, body, params, userAgent }) => {
-
     const dataset = await Datasets.get(params.id, params.name, true).then(getOrNotFound);
 
     await auth.canOrReject('entity.create', dataset);
 
     const properties = await Datasets.getProperties(dataset.id);
 
-    const partial = await Entity.fromJson(body, properties, dataset);
-
-    const sourceId = await Entities.createSource();
-    const entity = await Entities.createNew(dataset, partial, null, sourceId, userAgent);
+    // Destructure list of new entities and source if bulk operation
+    const { entities, source } = body;
+
+    if (!entities) {
+      // not a bulk operation
+      const partial = await Entity.fromJson(body, properties, dataset);
+      const sourceId = await Entities.createSource();
+      const entity = await Entities.createNew(dataset, partial, null, sourceId, userAgent);
+      // Entities.createNew doesn't return enough information for a full response so re-fetch.
+      return Entities.getById(dataset.id, entity.uuid).then(getOrNotFound);
+    } else {
+      // bulk operation
+      // TODO: maybe validate source fields
```

Comment:
I think we may need to revisit this (how the source of a bulk action is represented, if/how it should be validated) when we get to this issue: https://github.com/getodk/central/issues/583